File: Algo/Leetcode/134GasStation.cpp
//Run one through.
class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        int sum = 0;
        int total = 0;
        int j = -1;
        for(int i=0;i<gas.size();i++){
            sum += gas[i] - cost[i];
            total += gas[i]-cost[i];
            if(sum<0){
                sum = 0;
                j=i;
            }
        }
        
        return total>=0? j+1:-1;
    }
};

// more detail to understand
class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        int n = gas.size();
        int i = 0;
        while(i < n){               // start from every i
            int sumGas = 0;
            int sumCost = 0;
            int count = 0;
            while(count < n){        // if count == n, satisfied
                int j = (i + count) % n;
                sumGas += gas[j];
                sumCost += cost[j];
                if(sumCost > sumGas) break;   // at any point, gas if not enough for cost
                count++;
            }
            if(count == n){
                return i;
            } else {
                i = i + count + 1;             // if you can not go from i to j, you can not go from any point in (i,j) to j. You must start from new point
            }
        }
        return -1;
    }
};

File: Algo/Leetcode/134GasStation.py
class Solution(object):
    def canCompleteCircuit(self, gas, cost):
        """
        :type gas: List[int]
        :type cost: List[int]
        :rtype: int
        """
        n = len(gas)
        val = 0
        start = 0
        res = 0     # keep all the difference sum. If sum(diff) > 0, can find a way
        for i in range(n):
            val += (gas[i] - cost[i])
            res += (gas[i] - cost[i])
            # start can not be i
            if (val < 0):
                start = i+1
                val = 0

        return -1 if res < 0 else start